### PR TITLE
[COST-545]: Obtain credentials for upload to cloud.

### PR DESCRIPTION
**Pull request by chambridge**
* Collect pull-secret for token auth
* Collect user,password for basic auth

Related to:
https://issues.redhat.com/browse/COST-545

### CODE REVIEW

**Comment by chambridge** [comment]
Worked through various test flows:
- token get pull secret
- basic with no auth secret name
- basic with configured auth secret

Currently getting a docker build error, but not sure if its just a network issue:
```
 => CACHED [builder 4/10] COPY go.sum go.sum                                                                                                                                                                             0.0s
 => ERROR [builder 5/10] RUN go mod download                                                                                                                                                                             1.9s
------
 > [builder 5/10] RUN go mod download:
#11 1.875 go: github.com/openshift/library-go@v0.0.0-20200120153246-906409ae5e38 requires
#11 1.875 	bitbucket.org/ww/goautoneg@v0.0.0-20120707110453-75cd24fc2f2c: reading https://api.bitbucket.org/2.0/repositories/ww/goautoneg?fields=scm: 404 Not Found
------
failed to solve with frontend dockerfile.v0: failed to build LLB: executor failed running [/bin/sh -c go mod download]: runc did not terminate sucessfully
make: *** [docker-build] Error 1
```

**Comment by chambridge** [comment]
Fixed the docker build.

File: controllers/costmanagement_controller.go [hunk 1]
```diff
@@ -132,6 +158,94 @@ func GetClusterID(r *CostManagementReconciler, cost *costmgmtv1alpha1.CostManage
...
+	ctx := context.Background()
+	log := r.Log.WithValues("costmanagement", "GetPullSecretToken")
+	sl := &corev1.SecretList{}
+	err := r.List(
+		ctx,
+		sl,
+		&client.ListOptions{Namespace: openShiftConfigNamespace},
+	)
+	if err != nil {
+		if errors.IsNotFound(err) {
```
Comment (boaz0) [inline controllers/costmanagement_controller.go:171]: maybe `swtich-case` will be more readable:

```go
switch {
case errors.IsNotFound(err):
  ....
case errors.IsForbidden(err):
....
default:
....
}
```

**Review by maskarb — COMMENTED** [review]
Initial round of review:

File: controllers/costmanagement_controller.go [hunk 2]
```diff
@@ -132,6 +158,94 @@ func GetClusterID(r *CostManagementReconciler, cost *costmgmtv1alpha1.CostManage
...
+		return err
+	}
+	for _, s := range sl.Items {
+		if strings.Contains(s.Name, pullSecretName) {
+			encodedPullSecret := s.Data[pullSecretDataKey]
+			if len(encodedPullSecret) > 0 {
+				var pullSecret serializedAuthMap
+				if err := json.Unmarshal(encodedPullSecret, &pullSecret); err != nil {
+					log.Error(err, "Unable to unmarshal cluster pull-secret.")
+				}
```
[the remark this answers is not in the document]
Comment (chambridge) [reply controllers/costmanagement_controller.go:189]: yeah.updating.

File: controllers/costmanagement_controller.go [hunk 3]
```diff
@@ -132,6 +158,94 @@ func GetClusterID(r *CostManagementReconciler, cost *costmgmtv1alpha1.CostManage
...
+				if auth, ok := pullSecret.Auths[pullSecretAuthKey]; ok {
+					token := strings.TrimSpace(auth.Auth)
+					if strings.Contains(token, "\n") || strings.Contains(token, "\r") {
+						return fmt.Errorf("Cluster authorization token is not valid: contains newlines.")
+					}
+					if len(token) > 0 {
+						log.Info("Found cloud.openshift.com token")
+						costInput.BearerTokenString = token
+					} else {
```
[the remark this answers is not in the document]
Comment (chambridge) [reply controllers/costmanagement_controller.go:198]: sure. no need to keep looping. Will update.

File: controllers/costmanagement_controller.go [hunk 4]
```diff
@@ -132,6 +158,94 @@ func GetClusterID(r *CostManagementReconciler, cost *costmgmtv1alpha1.CostManage
...
+		} else if errors.IsForbidden(err) {
+			log.Info("Operator does not have permission to check pull-secret: %v", err)
+			err = nil
+		} else {
+			err = fmt.Errorf("Could not check pull-secret: %v", err)
+		}
+		return err
+	}
+	for _, s := range sl.Items {
+		if strings.Contains(s.Name, pullSecretName) {
```
[the remark this answers is not in the document]
Comment (chambridge) [reply controllers/costmanagement_controller.go:183]: I've flattened it out a bit and added additional case logging.

File: controllers/costmanagement_controller.go (hunk 1 above)
Comment (chambridge) [reply controllers/costmanagement_controller.go]: Done

File: controllers/costmanagement_controller.go [hunk 5]
```diff
@@ -132,12 +158,104 @@ func GetClusterID(r *CostManagementReconciler, cost *costmgmtv1alpha1.CostManage
...
+	log.Info("Secret namespace", "namespace", reqNamespace.Namespace)
+	secret := &corev1.Secret{}
+	namespace := types.NamespacedName{
+		Namespace: reqNamespace.Namespace,
+		Name:      costInput.AuthenticationSecretName}
+	err := r.Get(ctx, namespace, secret)
+	if err != nil {
+		switch {
+		case errors.IsNotFound(err):
+			log.Error(err, "Secret does not exist.")
```
Comment (abaiken) [inline controllers/costmanagement_controller.go:225]: Logging the error here still shows the traceback logs - but I'll leave it up to you whether you want to quiet them all or not.

**Review by abaiken — APPROVED** [review]
Left a comment about quieting logs but it isn't a blocker. I tested this with the default token and basic auth (both configured correctly & misconfigured). Great work 👍

File: controllers/costmanagement_controller.go (hunk 5 above)
Comment (chambridge) [reply controllers/costmanagement_controller.go:225]: I was going to leave the initial. You all can circle back if you prefer to quiet them all.

**Comment by boaz0** [comment]
LGTM :+1:

**Pull request merged**